ra-optimisation/search-tree-rao/src/main/java/com/powsybl/openrao/searchtreerao/commons/objectivefunctionevaluator/ObjectiveFunction.java

```diff
@@ -107,11 +114,18 @@ public ObjectiveFunction build(Set<FlowCnec> flowCnecs,
             // Unoptimized cnecs in operatorsNotToOptimizeInCurative countries
             if (raoParameters.getNotOptimizedCnecsParameters().getDoNotOptimizeCurativeCnecsForTsosWithoutCras()
                 && !operatorsNotToOptimizeInCurative.isEmpty()) {
-
-                this.withFunctionalCostEvaluator(new MinMarginEvaluator(flowCnecs, raoParameters.getObjectiveFunctionParameters().getType().getUnit(),
-                    new MarginEvaluatorWithMarginDecreaseUnoptimizedCnecs(marginEvaluator, operatorsNotToOptimizeInCurative, prePerimeterFlowResult)));
+                if (raoParameters.getObjectiveFunctionParameters().getType().costOptimization()) {
+                    addEvaluatorsForCostlyOptimization(flowCnecs, raoParameters, optimizedStates, marginEvaluator);
+                } else {
+                    this.withFunctionalCostEvaluator(new MinMarginEvaluator(flowCnecs, raoParameters.getObjectiveFunctionParameters().getType().getUnit(),
+                        new MarginEvaluatorWithMarginDecreaseUnoptimizedCnecs(marginEvaluator, operatorsNotToOptimizeInCurative, prePerimeterFlowResult)));
+                }
             } else {
-                this.withFunctionalCostEvaluator(new MinMarginEvaluator(flowCnecs, raoParameters.getObjectiveFunctionParameters().getType().getUnit(), marginEvaluator));
+                if (raoParameters.getObjectiveFunctionParameters().getType().costOptimization()) {
+                    addEvaluatorsForCostlyOptimization(flowCnecs, raoParameters, optimizedStates, marginEvaluator);
+                } else {
+                    this.withFunctionalCostEvaluator(new MinMarginEvaluator(flowCnecs, raoParameters.getObjectiveFunctionParameters().getType().getUnit(), marginEvaluator));
+                }
             }
```

Comment:
it would be cleaner this way I think
if (not optimized...) {
   marginEvaluator = new MarginEvaluatorWithMarginDecreaseUnoptimizedCnecs(...)
}
if (costOptimization()) {
   ...
} else {
   ...
}


It will also allow us to use the not optimized cnecs in combination with the costly optimization